oak_attestation_verification/src/verifier.rs

```diff
@@ -58,9 +59,28 @@ use crate::{
 // We don't use additional authenticated data.
 const ADDITIONAL_DATA: &[u8] = b"";
 
+pub trait AttestationVerifier {
+    fn verify(
+        &self,
+        evidence: &Evidence,
+        endorsements: &Endorsements,
+    ) -> anyhow::Result<DiceChainResult>;
+}
+
+/// Verifier that doesn't check the Evidence against Reference Values and only checks the DICE chain
+/// correctness.
+/// Should be only used for testing.
+pub struct InsecureAttestationVerifier;
+
+impl AttestationVerifier for InsecureAttestationVerifier {
+    fn verify(&self, evidence: &Evidence, _: &Endorsements) -> anyhow::Result<DiceChainResult> {
+        verify_dice_chain(evidence).context("couldn't verify the DICE chain")
```

Comment:
Why do we need AttestationVerifier, InsecureAttestationVerifier - couldn't we just expose the function, i.e. a variant of verify which doesn't have the reference_values parameter? My idea so far was to use only static functions in the interface.

Also, I have no idea where a SecureAttestationVerifier would fit into the pattern.

How about this:

pub fn verify_minimal(
    now_utc_millis: i64,
    evidence: &Evidence,
    endorsements: &Endorsements, 
) -> anyhow::Result < DiceChainResult > {
  verify_dice_chain(evidence)
}

This is a shortcut to verify which behaves identically to an actual verification, but with assumed (but not existent) ReferenceValues skipping over every single dispensable step. This would be much clearer IMO.

Reply:
Currently our client library accepts an implementation of the `AttestationVerifier`

And `InsecureAttestationVerifier` is just one of those implementations. If we provide a `verify_minimal` we would have to hardcode it into the client code.

The `SecureAttestationVerifier` will be just another implementation of the `AttestationVerifier` which may look like this:
```rust
struct SecureAttestationVerifier {
    reference_values: ReferenceValues,
}

impl SecureAttestationVerifier {
    fn create(reference_values: ReferenceValues) -> Self {...}
}

impl AttestationVerifier for SecureAttestationVerifier {
    fn verify(evidence, endorsements) -> ...
}
```

Reply:
This interface also allows us to create specific verifiers (i.e. `OakFunctionsAttestationVerifier`) which will create reference values during the `create` method. So that our prod clients won't need to construct reference values themselves (if they want to rely of the default ones).

Reply:
// Use only for testing, not in production.
pub fn verify_minimal(
  evidence: &Evidence,
) -> anyhow::Result < DiceChainResult > {
  verify_dice_chain(evidence)
}

(Alternative to the above)

Reply:
> couldn't we just expose the function, i.e. a variant of verify which doesn't have the reference_values parameter? My idea so far was to use only static functions in the interface.

This function will never be seen by our prod teams - they will only see the `OakClient` class that we provide to them. So it's `OakClient` that will need to either use the minimal verification for testing or use actual verification.

Reply:
I see. This seems like an intermediate Oak-Client specific layer. Is the solution to put into a separate module? Either here, or in oak_client? I noticed that oak_client has a verifier.rs module as well. If things go well this would cut a few direct dependencies on oak_attestation_verification, i.e. crates would only depend on oak_client, not on both. A layered approach. WDYT?

Reply:
Yes, this interface is just there for prod teams' convenience - i.e. it a part of the client SDK and doesn't have to the the core logic of the attestation library.

I can move it into the `oak_client` crate. And make `InsecureAttestationVerifier` call the `verify_minimal` function.

Reply:
Though we would still need to implement other verifiers (i.e. `CustomAttestationVerifier` that accepts any reference values, or application specific verifiers like `OakFunctionsAttestationVerifier`).

Where should we put those?

Reply:
Ideally they would belong to that module containing the present AttestationVerifier Rust structs, with all of its impl. However, the "custom" part seems to be the addition of a ReferenceValues proto instance which will always be hard-wired. That should be contained in the "custom" place, i.e. OakFunctionsAttestationVerifier would not exist (or it would exist in private somewhere in oak_functions_*.)  OakFunctionsAttestationVerifier (if it needs to be named) would just be the CustomAttestationVerifier and add its static instance of ReferenceValues.

Assuming that CustomAttestationVerifier and SecureAttestationVerifier are the same thing.

For C++ and Java: the JNI interface would just connect/implement the static verify functions (maybe some variants), so the wrappers would be reimplemented in the respective language.

Reply:
Moved `InsecureAttestationVerifier` to `oak_client` crate. But the `AttestationVerifier` is currently still in the `oak_attestation_verification`

Reply:
Update: moved the trait to `oak_client` as well